product_lifecycle/tests/test_product.py

```diff
@@ -30,3 +32,49 @@ def test_02_so_suggestion(self):
         self.assertTrue(so_line.show_replacement)
         widget_info = json.loads(so_line.replacement_info_widget)
         self.assertEqual(widget_info['product_id'], self.good_replacement.id)
+        # No show widget
+        order_form = Form(self.env['sale.order'])
+        order_form.partner_id = self.customer
+        with order_form.order_line.new() as line:
+            line.product_id = self.good_replacement
+        sale_order = order_form.save()
+        so_line = sale_order.order_line[0]
+        self.assertFalse(so_line.show_replacement)
+
+    def test_03_product_write(self):
+        self.doble_obsolete_product.write({'lifecycle_state': 'end'})
+        self.assertEqual(self.doble_obsolete_product.lifecycle_state, 'obsolete')
+        self.end_product.write({'lifecycle_state': 'obsolete'})
+        self.assertEqual(self.end_product.lifecycle_state, 'end')
+
+    def test_04_update_product_state(self):
+        # confirm picking for product in end
+        pickings = self.env['stock.picking'].search([
+            ('move_lines.product_id', '=', self.end_product.id),
+            ('state', 'not in', ('done', 'draft', 'cancelled'))])
+        ready = pickings.filtered(lambda p: p.state == 'assigned')
+        for pick in ready:
```

Comment:
Tests shouldn't include loops, because you are intended to know exactly what records you're diling with.

Reply:
any proposal?

Or I can just remove this test.